### ISSUE TITLE
Does UVM Tensor support autograd?

### ISSUE THREAD

**Issue by zxgx** [pytorch/FBGEMM#1189]
Firstly, this is definitely one of the coolest work I've ever seen in deep learning! 
Yet, I have got a lot of questions about `fbgemm_gpu`.

I notice that `SplitTableBatchedEmbeddingBagsCodegen` initializes embedding shards' weight as a registered buffer instead of nn.Parameter. Why should it be a registered buffer? Is it feasible to create a `DenseTableBatchedEmbeddingBagsCodegen` like module, which preserves the UVM & cache mechanisms while removes the fused optimizer part?

Generally, I want to know whether it is possible to init a UVM Tensor with `fbgemm.new_managed_tensor` as an ordinary CUDA Tensor, so that it can be used nn.Parameter with autograd engine?
If so, does this suffer any performance loss?

**Comment by zxgx** [comment]
Hi! I've written a demo script to explore above questions. Here is the code:
```python
import torch
import fbgemm_gpu
# I suppose I must import fbgemm_gpu to load ops:
# https://github.com/pytorch/FBGEMM/blob/ffc05b38ba48d2cc03a67f37ce3b09a93763ce54/fbgemm_gpu/fbgemm_gpu/__init__.py#L13
import psutil


fraction = 0.1  # 80 * 0.1 = 8GB cuda memory
device = torch.device("cuda", 1)
torch.cuda.set_per_process_memory_fraction(fraction, device=device)

total_mem = 16 * 1024 ** 3  # Large Tensor of size 16GB
dim = 128
row = total_mem // dim // 4
size = (row, dim)
dtype = torch.float32

fbgemm_tensor = torch.zeros(
    size,
    out=torch.ops.fbgemm.new_managed_tensor(torch.zeros(1, device=device, dtype=dtype), size),
    requires_grad=True
)

print(f"Tensor shape: {fbgemm_tensor.shape}, device: {fbgemm_tensor.device}, dtype: {fbgemm_tensor.dtype}")
print(f"CUDA Memory allocated: {torch.cuda.memory_allocated(device)}, reserved: {torch.cuda.memory_reserved(device)}")
print(f"CPU Memory retrieved by psutils: {psutil.Process().memory_info().rss / 1024**3}")

try:
    torch_tensor = torch.randn(size, device=device, dtype=dtype)
    print("Successfully init torch tensor")
except Exception:
    print(f"Failed to init torch tensor")


size = (3, 4)
autograd_tensor = torch.randn(
    size,
    out=torch.ops.fbgemm.new_managed_tensor(torch.zeros(1, dtype=dtype, device=device), size),
    requires_grad=True
)
print(f"test tensor: {autograd_tensor}")

scalar = autograd_tensor.sum()
scalar.backward()
print(f"leaf tensor: {scalar.item()}, grad: {autograd_tensor.grad}")

scalar = (fbgemm_tensor / row).sum()
scalar.backward()
print(f"leaf tensor: {scalar.item()}, grad: {fbgemm_tensor.grad}")

```

The result is shown as the following screenshot:
![image](https://user-images.githubusercontent.com/34452939/177671823-d5c73361-0a44-4c3d-8b03-f95e0ec9bfac.png)

So I suppose the computation results for UVM Tensors are allocated as CUDA Tensors. Is that true?
How can I keep the computation results as UVM Tensors?

btw, since both `torch.cuda.memory_xxx` and `psutils` couldn't get the correct statistics, how can I retrieve the memory size allocated for the UVM Tensor?

**Comment by jianyuh** [comment]
Thanks for your interests using FBGEMM!

> `SplitTableBatchedEmbeddingBagsCodegen` initializes embedding shards' weight as a registered buffer instead of nn.Parameter. Why should it be a registered buffer? 

IIRC `register_buffer` will present  from updating the embedding weights using SGD mechanism (https://stackoverflow.com/a/57541778), where as you noticed, we have the fused optimizer.

> Is it feasible to create a `DenseTableBatchedEmbeddingBagsCodegen` like module, which preserves the UVM & cache mechanisms while removes the fused optimizer part?

Yes: feel free to submit a PR if you are interested. Currently we mainly use dense table batched embedding bag (TBE) for small embedding tables (using data parallelism) so there is no use case on our side to use UVM & cache for dense TBE. Potentially we could create dense TBE + UVM w/o fused optimizer, but it is not in our priority list now.

**Comment by zxgx** [comment]
As you can see in the above example, when I create a tensor larger than the GPU memory, the UVM makes it easy to hold across device and host. 
However, it seems that the computation results are allocated on GPU side, so I suppose I also need to write c++/cuda code to directly manipulate the unified memory.  Is that right?

**Comment by jianyuh** [comment]
Looks like so. Currently PyTorch core doesn't support UVM tensor, so the autograd tensor will be GPU device tensor only.

BTW, we have the API to check whether a tensor is UVM tensor or not in FBGEMM: `torch.ops.fbgemm.is_uvm_tensor`. e.g., https://github.com/pytorch/FBGEMM/blob/c1a63cb6b9e3f369619ecd53f7aadf1a6d9384f3/fbgemm_gpu/test/uvm_test.py#L49

**Comment by jianyuh** [comment]
cc @ngimel on the potential interests to upstream UVM tensor support into PyTorch Core.

**Comment by zxgx** [comment]
Great, thanks for your help!
Hope UVM could be a new feature in pytorch.
